video/out/vo_dmabuf_wayland.c

```diff
@@ -0,0 +1,357 @@
+/*
+ * Based on vo_gl.c by Reimar Doeffinger.
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <unistd.h>
+
+#include "config.h"
+
+#ifdef HAVE_VAAPI
+#include <va/va_drmcommon.h>
+#include <libavutil/hwcontext_drm.h>
+#endif
+
+#include "mpv_talloc.h"
+#include "common/global.h"
+#include "vo.h"
+#include "video/mp_image.h"
+
+#include "gpu/hwdec.h"
+#include "gpu/video.h"
+
+#include "video/vaapi.h"
+#include "present_sync.h"
+#include "wayland_common.h"
+#include "generated/wayland/linux-dmabuf-unstable-v1.h"
+#include "generated/wayland/viewporter.h"
+#include "wlbuf_pool.h"
+
+struct priv {
+    struct mp_log *log;
+    struct ra_ctx *ctx;
+    struct mpv_global *global;
+    struct ra_hwdec_ctx hwdec_ctx;
+    int events;
+
+    struct wl_shm_pool *solid_buffer_pool;
+    struct wl_buffer *solid_buffer;
+    struct wlbuf_pool *wlbuf_pool;
+    bool want_reset;
+    uint64_t reset_count;
+
+#ifdef HAVE_VAAPI
+    VADisplay display;
+#endif
+};
+
+#ifdef HAVE_VAAPI
+static uintptr_t vaapi_key_provider(struct mp_image *src){
+    return va_surface_id(src);
+}
+
+/* va-api dmabuf importer */
+static bool vaapi_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_entry* entry,
+                                                        struct zwp_linux_buffer_params_v1 *params)
```

Comment:
Github's webcrap seems to mangle the horizontal spacing so it looks deceiving. According to my editors, these should be:

```suggestion
static bool vaapi_dmabuf_importer(struct mp_image *src, struct wlbuf_pool_entry* entry,
                                  struct zwp_linux_buffer_params_v1 *params)
```